Adds ShaderMaterial properties to trigger inclusion of extensions

liammagee opened the pull request:
This change implements #7379 by adding three further properties to ShaderMaterial, to trigger inclusion of GL extensions (EXT_frag_depth, WEBGL_draw_buffers, EXT_shader_texture_lod). It follows the pattern of:

```
material.derivatives = true; 
```

liammagee committed:
Adds fragDepth, drawBuffers and shaderTextureLOD properties to ShaderMaterial

mrdoob commented:
I think this is a bit dirty... How about...

```javascript
this.extensions = {
 	derivatives: false, // set to use derivatives
	fragDepth: false, // set to use fragment depth values
	drawBuffers: false, // set to use draw buffers
	shaderTextureLOD: false // set to use shader texture LOD
}
```

We'll also need a getter/setter for backwards compatibility:

```javascript
Object.defineProperties( THREE.ShaderMaterial.prototype, {
	derivatives: {
		get: function () {
			return this.extensions.derivatives;
		},
		set: function ( value ) {
			console.warn( 'THREE. ShaderMaterial: .derivatives has been moved to .extensions.derivatives.' );
			this.extensions.derivatives = value;
		}
	}
} );

liammagee committed:
Updates ShaderMaterial with extensions property

liammagee commented:
I've added that in, but only ShaderMaterial has this extensions property, which caused problems in the fragment generation. I've added a generateExtensions() method modelled on generateDefines() to try to handle this in an elegant way.

mrdoob commented on src/renderers/webgl/WebGLProgram.js:
```diff
@@ -2,6 +2,21 @@ THREE.WebGLProgram = ( function () {
+		var chunks = [
+			( parameters.logarithmicDepthBuffer || extensions.derivatives ) && rendererExtensions.get( 'EXT_frag_depth' ) ? '#extension GL_OES_standard_derivatives : enable' : '',
+			( extensions.fragDepth ) && rendererExtensions.get( 'EXT_frag_depth' ) ? '#extension GL_EXT_frag_depth : enable' : '',
+			( extensions.drawBuffers ) && rendererExtensions.get( 'WEBGL_draw_buffers' ) ? '#extension GL_EXT_draw_buffers : enable' : '',
```
I think this one uses "require" instead of "enable"?

liammagee committed:
Changes GL_EXT_draw_buffers enable to required

liammagee committed:
Fix typo

liammagee commented:
I've fixed GL_EXT_draw_buffers to use 'require'.

mrdoob commented:
Thanks!
